File: scripts/tasks/verify_config_backup.py
"""Verify encrypted config backup can be decrypted and contains required paths."""

from __future__ import annotations

import os
import shutil
import tempfile
from pathlib import Path

from scripts.common import log_step, run_or_throw
from scripts.crypto import CryptoFormat, decrypt

CONFIG_FORMAT = CryptoFormat(magic="SCRYCFG1")

REQUIRED_RESTORE_PATHS = [
    ".openclaw/openclaw.json",
    ".openclaw/credentials",
    ".openclaw/cron/jobs.json",
    ".openclaw/identity",
]


def verify_config_backup() -> None:
    passphrase = os.environ.get("SCRY_CONFIG_BACKUP_PASSPHRASE", "")
    repo_root = Path.cwd().resolve()
    encrypted_file = Path(
        os.environ.get(
            "SCRY_CONFIG_BACKUP_FILE",
            str(repo_root / "vault" / "config" / "critical-configs.tar.enc"),
        )
    )

    log_step("Checking config backup verification prerequisites")

    if len(passphrase) < 16:
        raise RuntimeError(
            "Set SCRY_CONFIG_BACKUP_PASSPHRASE with at least 16 characters before verification."
        )

    if not encrypted_file.exists():
        raise RuntimeError(f"Encrypted backup file not found: {encrypted_file}")

    encrypted = encrypted_file.read_bytes()

    log_step("Decrypting and extracting backup payload to temp workspace")

    temp_dir = Path(tempfile.mkdtemp(prefix="scry-config-verify-"))

    try:
        plaintext = decrypt(encrypted, passphrase, CONFIG_FORMAT)
        tar_path = temp_dir / "critical-configs.tar"
        tar_path.write_bytes(plaintext)

        run_or_throw(["tar", "-xf", str(tar_path), "-C", str(temp_dir)])

        log_step("Validating required restore paths")
        missing: list[str] = []

        for rel_path in REQUIRED_RESTORE_PATHS:
            abs_path = temp_dir / rel_path
            if not abs_path.exists():
                missing.append(rel_path)

        if missing:
            raise RuntimeError(
                f"Restore preview missing required paths: {', '.join(missing)}"
            )

        tar_size = tar_path.stat().st_size

        log_step("Config backup verification passed")
        print(f"artifact: {encrypted_file}")
        print(f"decrypted tar bytes: {tar_size}")
        print(f"required restore paths: {len(REQUIRED_RESTORE_PATHS)} present")
        print(f"home reference: {os.environ.get('HOME', str(Path.home()))}")
    finally:
        shutil.rmtree(temp_dir, ignore_errors=True)

File: packages/cli/src/commands/verify-config-backup.ts
import { createHash } from 'node:crypto'
import fs from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'

import {
  ConfigBackupMetadataSchema,
  decryptPayload,
  formatZodError,
  logStep,
  runCommand,
} from '@scry-home/core'

const configFormat = { magic: 'SCRYCFG1' } as const

export const verifyConfigBackup = async () => {
  const passphrase = process.env.SCRY_CONFIG_BACKUP_PASSPHRASE ?? ''
  const repoRoot = process.cwd()
  const encryptedFile =
    process.env.SCRY_CONFIG_BACKUP_FILE ??
    path.join(repoRoot, 'vault', 'config', 'critical-configs.tar.enc')
  const metadataFile =
    process.env.SCRY_CONFIG_METADATA_FILE ??
    path.join(repoRoot, 'vault', 'config', 'critical-configs.meta.json')

  logStep('Checking config backup verification prerequisites')

  if (passphrase.length < 16) {
    throw new Error(
      'Set SCRY_CONFIG_BACKUP_PASSPHRASE with at least 16 characters before verification.',
    )
  }

  const encrypted = await fs.readFile(encryptedFile)
  const metadataResult = ConfigBackupMetadataSchema.safeParse(
    JSON.parse(await fs.readFile(metadataFile, 'utf8')),
  )

  if (!metadataResult.success) {
    throw new Error(`Config backup metadata is invalid: ${formatZodError(metadataResult.error)}`)
  }

  const metadata = metadataResult.data

  if (metadata.encryptedBackupSha256) {
    const actualSha = createHash('sha256').update(encrypted).digest('hex')
    if (actualSha !== metadata.encryptedBackupSha256) {
      throw new Error(
        `Encrypted backup SHA-256 does not match metadata. metadata=${metadata.encryptedBackupSha256} actual=${actualSha}`,
      )
    }
  }

  logStep('Decrypting and extracting backup payload to temp workspace')
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'scry-config-verify-'))

  try {
    const plaintext = decryptPayload(encrypted, passphrase, configFormat)
    const tarPath = path.join(tempDir, 'critical-configs.tar')
    await fs.writeFile(tarPath, plaintext)
    await runCommand(['tar', '-xf', tarPath, '-C', tempDir])

    logStep('Validating recorded backup contents')
    const missing = []
    for (const relativePath of metadata.includedPaths) {
      const absolutePath = path.join(tempDir, relativePath)
      const exists = await fs
        .access(absolutePath)
        .then(() => true)
        .catch(() => false)
      if (!exists) {
        missing.push(relativePath)
      }
    }

    if (missing.length > 0) {
      throw new Error(`Restore preview missing recorded paths: ${missing.join(', ')}`)
    }

    const tarSize = (await fs.stat(tarPath)).size
    logStep('Config backup verification passed')
    process.stdout.write(`artifact: ${encryptedFile}\n`)
    process.stdout.write(`metadata: ${metadataFile}\n`)
    process.stdout.write(`decrypted tar bytes: ${tarSize}\n`)
    process.stdout.write(`recorded backup paths: ${metadata.includedPaths.length} present\n`)
    process.stdout.write(`source fingerprint: ${metadata.sourceFingerprint ?? 'unknown'}\n`)
    process.stdout.write(`home reference: ${process.env.HOME ?? os.homedir()}\n`)
  } finally {
    await fs.rm(tempDir, { force: true, recursive: true })
  }
}
